fn/async.go

```diff
@@ -0,0 +1,76 @@
+package fn
+
+import "errors"
+
+type Async[A any] interface {
+	Await() (*A, error)
+	Cancel()
+}
+
+var (
+	ContextCancelled = errors.New("context cancelled")
+	AsyncCancelled   = errors.New("async cancelled")
+)
+
+// RunAsyncWithCancel is a function that will transform a synchronous function
+// and make it a cancellable asynchronous call. This can be useful if the
+// function in the argument can block indefinitely and the calling context needs
+// a way to abandon waiting for it to complete.
+func RunAsyncWithCancel[A any](
+	f func() (*A, error),
+	contextCancel <-chan struct{},
+) Async[A] {
+	a := async[A]{
+		contextCancel: contextCancel,
+		targetCancel:  make(chan struct{}),
+		res:           make(chan *A),
+		err:           make(chan error),
+	}
+	go func() {
+		res, err := f()
```

Comment:
should we allow the function to take in a `context.Context` so that we can cancel it if `Cancel` is called?

Reply:
Yeah I think this is a good consideration. I don't have much experience using `context.Context`. The trick would be that we would need to make the argument function accept a `Option[Context]`, or maybe add a second API that allows us to thread a function that accepts a context in.

Reply:
just change the current argument to `f func(ctx context.Context) (*A, error)`?

Reply:
I think `context.Context` can do a lot of the work for you here actually. Wdyt of this:

```
type Async[A any] interface {
	Await() (*A, error)
	Cancel()
}

func RunAsyncWithCancel[A any](ctx context.Context,
	f func(ctx context.Context) (*A, error)) Async[A] {

	ctxc, cancel := context.WithCancel(ctx)

	a := async[A]{
		ctx:    ctxc,
		res:    make(chan *A, 1),
		err:    make(chan error, 1),
		cancel: cancel,
	}

	a.wg.Add(1)
	go func() {
		defer a.wg.Done()

		res, err := f(ctxc)
		if err != nil {
			a.err <- err
			
			return
		}
		
		a.res <- res
	}()

	return &a
}

func RunAsync[A any](f func(ctx context.Context) (*A, error)) Async[A] {
	return RunAsyncWithCancel(context.Background(), f)
}

func (a *async[A]) Await() (*A, error) {
	select {
	case r := <-a.res:
		return r, nil
	case e := <-a.err:
		return nil, e
	case <-a.ctx.Done():
		return nil, a.ctx.Err()
	}
}

func (a *async[A]) Cancel() {
	a.cancel()

	a.wg.Wait()
}

type async[A any] struct {
	ctx    context.Context
	cancel func()
	res    chan *A
	err    chan error

	wg sync.WaitGroup
}
```

Reply:
Seems reasonable to me.